### ISSUE TITLE
2.0 Branch

### ISSUE THREAD

**Issue by TomBZombie** [TomBZombie/Dice#58]
I've just added a 2.0 branch with several changes. The main change is the use of arrays for configuration instead of the rule object as discussed in #56.

The other major change is structural one that gives a performance increase that actually makes Dice faster than containers that don't use reflection such as Pimple when a class is instantiated more than once by reducing the logic in each create() call.

This is a proof-of-concept more than anything but I think it's the right direction to go in.

**Comment by TomBZombie** [comment]
For 2.0, another change I think may be useful is rule inheritance, currently, when you apply a rule to something, it overwrites whatever rule exists for that class at the moment.

So at the moment:

```php
$rule = new \Dice\Rule();
$rule->shared = true;
$dice->addRule('A', $rule);


$rule = new \Dice\Rule();
$rule->substitutions['B'] = new \Dice\Instance('C');
$dice->addRule('A', $rule);
```

The rule for A gets overrriden by the second rule.  In this instance should A really be both shared and have the substitution set?

Where this gets particularly useful is inheritance. Consider:

```php
class A { }
class B extends A {}
```

At the moment, if I do this:

```php
$rule = new \Dice\Rule();
$rule->shared = true;
$dice->addRule('A', $rule);

var_dump($dice->getRule('B')->shared); // TRUE

$rule = new \Dice\Rule();
$rule->substitutions['C'] = new \Dice\Instance('D');
$dice->addRule('B', $rule);

var_dump($dice->getRule('B')->shared); // FALSE
```

`B` is not flagged as shared. I haven't explicitly said that B shouldn't be shared but it will be set to false because that is the default value in the rule.

A more intuitive approach may be to make rules inherit from the current so that this happens:


```php
$rule = new \Dice\Rule();
$rule->shared = true;
$dice->addRule('A', $rule);

var_dump($dice->getRule('B')->shared); // TRUE

$rule = new \Dice\Rule();
$rule->substitutions['C'] = new \Dice\Instance('D');
$dice->addRule('B', $rule);

var_dump($dice->getRule('B')->shared); // TRUE
```

Thoughts?

Using the new syntax this would be:

```php
$dice->addRule('A', ['shared' => true]);

var_dump($dice->getRule('B')['shared']) //TRUE

$dice->addRule('B', ['substitutions' => ['C' => 'D']);

var_dump($dice->getRule('B')['shared']) //TRUE
```

(Even looking at this example code makes me think the updated syntax is a huge improvement!)

CC @garrettw, you had a problem with this before so your opinion is very helpful here :)

**Comment by garrettw** [comment]
It's a great idea, but I get a gut feeling that there might be some edge cases where this would break down. I'll let you know if I can come up with a scenario; I can't at the moment.

I've already devised a different way of achieving this that is a bit more explicit: I added a parameter to `addRule()` that accepts another existing rule that you wish to merge with the one you're trying to add.
I'd love to hear what you think about it.
```php
    public function addRule($match, array $rule, $mergewith = '*')
    {
        $match = ltrim(strtolower($match), '\\');

        if ($mergewith !== '*'):
            $mergewith = ltrim(strtolower($mergewith), '\\');
        endif;

        if (!is_array($mergewith)):
            $mergewith = $this->rules[$mergewith];
        endif;

        $this->rules[$match] = array_merge($mergewith, $rule);
    }
```

**Comment by TomBZombie** [comment]
The problem with that approach is it always merges with * so if you applied 2 rules to the same class they wouldn't actually get merged, similarly it won't merge using subclasses.

If you ever did need to merge with anything other than what's already applied to the class you could always just get the rule and modify it as you can now.

**Comment by garrettw** [comment]
It only merges with * by default. If you want to apply two successive rules to the same class, you can either write it the old way (in which case the second rule would overwrite the first, same as the old behavior), or you can do:
`$dice->addRule('B', $rule, 'B');`
which would have the desired inheriting effect.

**Comment by TomBZombie** [comment]
It's a little more verbose as well. In my opinion, this is more intuitive:
```php
$dice->addRule('B', ['shared' => true]);
$dice->addRule('B', ['constructParams' => 'foo']);
```

Which makes B both shared and with 'foo' as a constructor parameter. What I'm trying to think of is whether there is a situation where this is not desirable. Even if you want to change the rule, you can always update the property you want to overwrite:

```php
$dice->addRule('B', ['shared' => true]);
$dice->addRule('B', ['constructParams' => 'foo', 'shared' => false]);
```

Although most of the time I don't think you'd need to do this.

**Comment by garrettw** [comment]
You're probably right, because the semantics of "add" imply a cumulative effect, as opposed to something like "set".

**Comment by garrettw** [comment]
Also could you take a look at the JSON and XML loaders? I noticed that they need some updating, not only for the new rule/instance format, but also a few other things.
Json::getComponent() doesn't take two params anymore, but lines 29 and 31 are still passing a second param. I assume the behavior is still as expected?
Xml::getComponent() hasn't been updated like its Json counterpart.
Can your JSON and XML formats be made to match the names of the array keys in Dice::$rules?

**Comment by TomBZombie** [comment]
I'll start work on the Xml/Json loaders. 

I've just added support for factory methods e.g. the equivalent of `$myfactory->create('foo');`:

```json
{
   "substitutions": {"instance": ['MyFactory', 'create'], "params": "foo"}
}
```

The params list gets expanded so you can use 

```json
{
   "substitutions": {"instance": ['MyFactory', 'create'], "params": {"instance": "$Whatever"}}
}
```

This removes the need for the Loader\Callback class, as this was the functionality it added that didn't already exist inside the Dice Class. 

The implementation is a bit messy and I need to add some tests.

**Comment by TomBZombie** [comment]
I've now updated the Json and XML loaders to pass all tests and match the names in XML. The only problem with this is that semantically, the XML loader now doesn't make much sense:

```xml
<rule>
    <name>Foo</name>
    <newInstances>A</newInstances>
    <newInstances>B</newInstances>
    <newInstances>C</newInstances>
</rule>
```

There's not really a nice way to represent an array inside XML without adding an extra level of nesting.

**Comment by garrettw** [comment]
So why not add the extra nesting? Seems like that would make plenty of sense.

**Comment by TomBZombie** [comment]
Unless anyone has any objections I think 2.0 is ready to go live once I've updated the documentation. I've added a v2.0-Transitional branch which contains a wrapper that lets you use the old syntax with the new version

**Comment by daniel-meister** [comment]
No objections just a question. Will v2.0 only target PHP 5.6 or is there a plan for a 5.4/5.5 compatibility branch as well?

**Comment by garrettw** [comment]
@daniel-meister good question. I know a lot of shared web hosts aren't on 5.6 yet, which is why I maintain 5.4/5.5 compatibility on my fork.

**Comment by TomBZombie** [comment]
I'll backport it to 5.4 compatibility. Ideally I'd just have the one branch but PHP5.6 introduced some rather nice features that makes dice significantly more efficient (`new $class(...$args))` instead of reflection being the main one!). I intend to keep compatibility with PHP versions that are officially supported by php.net

**Comment by TomBZombie** [comment]
I was a little uneasy about this not working quite as it looks like it should:

```php

$rule = ['shared' => true];

$dice->addRule('Foo', $rule);

$appliedRule  = $dice->getRule('Foo');

var_dump($rule === $appliedRule);

```

Logic would suggest that this should evaluate to true, however it wasn't due to having the defaults stored under * and being merged. I've changed the way this works.

It's not great, but once PHP7 is out and supports ?? it can be tidied up a lot.

**Comment by TomBZombie** [comment]
I've just released 2.0 which handles all the open issues and I've released a version for PHP5.4-5.5, see the relevant branch/release :)
